### PR TITLE
cmake: zephyr_library_amend feature

### PR DESCRIPTION
**Pull request by tejlmand**
# Summary

With module support in Zephyr: https://docs.zephyrproject.org/latest/guides/modules.html?highlight=modules#modules-external-projects

more code will be located out of tree.
This means that in some cases, Kconfig settings may be set to `yes` and thus create a zephyr_library, but no file is place in the lib.

As example, high level settings for a driver might result in a zephyr_library is selected for creation, but the code for the specific driver is located out-of-tree.

This can result in an empty CMake library with the following error:
```
CMake Error at /projects/github/ncs/zephyr/cmake/extensions.cmake:411 (add_library):                       
  No SOURCES given to target: drivers__entropy
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/extensions.cmake:388 (zephyr_library_named)
  /projects/github/ncs/zephyr/drivers/entropy/CMakeLists.txt:3 (zephyr_library)
```

This commit introduces the cmake extension `zephyr_library_amend` which offers
a generic way to add files from out-of-tree into a Zephyr library and still use the existing zephyr cmake extension functions.

This function allows for adding files in an out-of-tree Zephyr module
to a zephyr library created in zephyr repo CMake files.

As example:
`drivers/entropy/CMakeLists.txt` creates an zephyr library as:
```
zephyr_library()
```
only available to zephyr itself.

The amend function allows to amend to such a lib, by creating a
CMakeLists.txt file following identical folder structure in a Zephyr
Module: `<zephyr_module_oot>/drivers/entropy/CMakeLists.txt`
```
zephyr_library_amend()
zephyr_library_sources_if_kconfig(entropy_oot_driver.c) # Sources are amended to the original library
```

Requirement when using `zephyr_library_amend`.
1) If original zephyr `CMakeList.txt` is located in:
`${ZEPHYR_BASE}/drivers/entropy`
then the amending `CMakeLists.txt` must be located in
`<ZEPHYR_MODULE_DIR>/drivers/entropy`

This PR is related to: #14527

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -375,6 +375,47 @@ macro(zephyr_library_named name)
...
+# zephyr_library_amend()
+# zephyr_libray_add_sources(...)
+#
+macro(zephyr_library_amend)
+  # This is a macro because we need to ensure the ZEPHYR_CURRENT_LIBRARY and
+  # following zephyr_library_* calls are executed within the scope of the
+  # caller.
+  zephyr_library_amend_module_lib_name(lib_name)
+
+  set(ZEPHYR_CURRENT_LIBRARY ${IMAGE}${name})
```
Comment (SebastianBoe) [inline cmake/extensions.cmake]: Remove ${IMAGE}

Comment (SebastianBoe) [inline cmake/extensions.cmake]: ${lib_name}, not ${name}

Comment (tejlmand) [reply cmake/extensions.cmake]: Fixed

Comment (tejlmand) [reply cmake/extensions.cmake]: fixed

File: cmake/extensions.cmake [hunk 2]
```diff
@@ -375,6 +375,49 @@ macro(zephyr_library_named name)
...
+  file(RELATIVE_PATH name ${ZEPHYR_CURRENT_MODULE_DIR} ${CMAKE_CURRENT_LIST_FILE})
+
+  # Remove the filename (driver/serial/CMakeLists.txt => driver/serial)
+  get_filename_component(name ${name} DIRECTORY)
+
+  # Replace / with __ (driver/serial => driver__serial)
+  string(REGEX REPLACE "/" "__" name ${name})
+
+  set(${lib_name} ${name})
+endmacro()
```
Comment (SebastianBoe) [inline cmake/extensions.cmake]: Let's increase the code re-use with zephyr_library_get_current_dir_lib_name

Comment (tejlmand) [reply cmake/extensions.cmake]: agree, and done.

**Review by SebastianBoe — APPROVED** [review]

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -344,15 +344,15 @@ endmacro()
...
 
-# Determines what the current directory's lib name would be and writes
-# it to the argument "lib_name"
-macro(zephyr_library_get_current_dir_lib_name lib_name)
+# Determines what the current directory's lib name would be according to the
+# provided base and writes it to the argument "lib_name"
+macro(zephyr_library_get_current_dir_lib_name base lib_name)
   # Remove the prefix (/home/sebo/zephyr/driver/serial/CMakeLists.txt => driver/serial/CMakeLists.txt)
-  file(RELATIVE_PATH name ${ZEPHYR_BASE} ${CMAKE_CURRENT_LIST_FILE})
+  file(RELATIVE_PATH name ${base} ${CMAKE_CURRENT_LIST_FILE})
```
Comment (thomasstenersen) [inline cmake/extensions.cmake]: Do we need to check if `${CMAKE_CURRENT_LIST_FILE}` is in a sub-directory of `${base}`? I.e. no `..` in the resulting `name`.

**Review by thomasstenersen — COMMENTED** [review]
A minor comment, otherwise it looks good to me :+1:

File: cmake/extensions.cmake (hunk 3 above)
Comment (tejlmand) [reply cmake/extensions.cmake]: Must it be restricted to be in a sub-dir ? and in that case, why ?
I know the existing functionality creates the funny looking libs which contains `..` when using `zephyr_library` out-of-tree.

But this PR does not address that issue, and therefore it tries to keep the behavior as-is for all existing use of `zephyr_library`.

Comment (thomasstenersen) [reply cmake/extensions.cmake]: Not really, had a feeling `..`could cause some troubles. Nevermind it then :)

**Review by thomasstenersen — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**